### PR TITLE
[ie/GameDevTV] Add extractor

### PR DESCRIPTION
**Pull request by stratus-ss**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Adds support for gamedev.tv

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?

- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))

</details>

### CODE REVIEW

File: yt_dlp/extractor/gamedevtv.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    join_nonempty,
+    parse_iso8601,
+    str_or_none,
+    url_or_none,
+)
+from ..utils.traversal import traverse_obj
+
+
+class GameDevTVDashboardIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?gamedev\.tv/dashboard/courses/(?P<id>\d+)'
```
Comment (seproDev) [inline yt_dlp/extractor/gamedevtv.py:18]: Imo. the extractor should support individual videos as well.
Given that only the course api request is needed, I think a single extractor is fine. Just add an optional sub_id to the regex and, if present, only return a single video.

**Review by seproDev — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/gamedevtv.py (hunk 1 above)
Comment (stratus-ss) [reply yt_dlp/extractor/gamedevtv.py:18]: I am very open as to how that might be achieved but there really isnt a way for the user to know the IDs of a single video. You'd have to open up the dev tools in the web browser and examin the site json in order to determine that information

The UI is completely dynamic and generated on the fly in javascript. The URL does not change every. From the users' perspective, the player loads a playlist and you click the playlist without the web page ever reloading

Comment (seproDev) [reply yt_dlp/extractor/gamedevtv.py:18]: For me the URL does change:
`https://www.gamedev.tv/dashboard/courses/63` is the main course URL
and then clicking on the different lectures, the URL in the browser does change between:
```
https://www.gamedev.tv/dashboard/courses/63/2279
https://www.gamedev.tv/dashboard/courses/63/2280
https://www.gamedev.tv/dashboard/courses/63/2281
https://www.gamedev.tv/dashboard/courses/63/2282
https://www.gamedev.tv/dashboard/courses/63/2283
```
Those are the IDs of the different lectures. We can still just fetch the entire course (afaik. there is no API for the individual lectures) and then, if the lecture_id is passed, filter the playlist to only return a single result with the matching id.

Comment (stratus-ss) [reply yt_dlp/extractor/gamedevtv.py:18]: Thats interesting, the entire time i was working on it, the website did not exhibit this behaviour.

I'll see what I can come up with to accommodate this request

Comment (stratus-ss) [reply yt_dlp/extractor/gamedevtv.py:18]: The regex probably looks like this `r"https?://(?:www\.)?gamedev\.tv/dashboard/courses/(?P<id>\d+)/?(?P<sub_id>\d*)"`

I am struggling to see how I would implement your suggestion however. _match_id() will get the course id, but I didn't see anything that would help in this case. 

I was considering filtering the data dict and removing everything before the object is traversed but i suspect I am missing some built in facility that yt-dlp has

Comment (seproDev) [inline yt_dlp/extractor/gamedevtv.py:18]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?gamedev\.tv/dashboard/courses/(?P<course_id>\d+)(?:/(?P<lecture_id>\d+))?'
```

File: yt_dlp/extractor/gamedevtv.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                        'episode_number': ('order', {int_or_none}),
+                        'duration': ('video', 'duration_in_sec', {int_or_none}),
+                        'timestamp': ('video', 'created_at', {parse_iso8601}),
+                        'modified_timestamp': ('video', 'updated_at', {parse_iso8601}),
+                        'thumbnail': ('video', 'thumbnailUrl', {url_or_none}),
+                    }),
+                }
+
+    def _real_extract(self, url):
+        course_id = self._match_id(url)
```
Comment (seproDev) [inline yt_dlp/extractor/gamedevtv.py:94]:
```suggestion
        course_id, lecture_id = self._match_valid_url(url).group('course_id', 'lecture_id')
```

File: yt_dlp/extractor/gamedevtv.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        course_info = traverse_obj(data, {
+            'title': ('title', {str}),
+            'tags': ('tags', ..., 'name', {str}),
+            'categories': ('categories', ..., 'title', {str}),
+            'timestamp': ('created_at', {parse_iso8601}),
+            'modified_timestamp': ('updated_at', {parse_iso8601}),
+            'thumbnail': ('image', {url_or_none}),
+        })
+
+        return self.playlist_result(self._entries(data, course_id, course_info), course_id, **course_info)
```
Comment (seproDev) [inline yt_dlp/extractor/gamedevtv.py:108]:
```suggestion
        entries = self._entries(data, course_id, course_info, lecture_id)
        if lecture_id:
            lecture = next(entries, None)
            if not lecture:
                raise ExtractorError('Lecture not found')
            return lecture

        return self.playlist_result(entries, course_id, **course_info)
```

File: yt_dlp/extractor/gamedevtv.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    def _entries(self, data, course_id, course_info):
+        for section in traverse_obj(data, ('sections', ..., {dict})):
+            section_info = traverse_obj(section, {
+                'season_id': ('id', {str_or_none}),
+                'season': ('title', {str}),
+                'season_number': ('order', {int_or_none}),
+            })
+            for lecture in traverse_obj(section, ('lectures', lambda _, v: url_or_none(v['video']['playListUrl']))):
+                display_id = join_nonempty(course_id, section_info.get('season_id'), lecture.get('id'))
```
Comment (seproDev) [inline yt_dlp/extractor/gamedevtv.py:67]:
```suggestion
                if lecture_id and str(lecture.get('id')) != lecture_id:
                    continue
                display_id = join_nonempty(course_id, section_info.get('season_id'), lecture.get('id'))
```

File: yt_dlp/extractor/gamedevtv.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            raise
+
+        self._API_HEADERS['Authorization'] = f'{response["token_type"]} {response["access_token"]}'
+
+    def _real_initialize(self):
+        if not self._API_HEADERS.get('Authorization'):
+            self.raise_login_required(
+                'This content is only available with purchase', method='password')
+
+    def _entries(self, data, course_id, course_info):
```
Comment (seproDev) [inline yt_dlp/extractor/gamedevtv.py:59]:
```suggestion
    def _entries(self, data, course_id, course_info, lecture_id):
```

**Review by seproDev — COMMENTED** [review]
This is what I would have done

File: yt_dlp/extractor/gamedevtv.py (hunk 2 above)
Comment (stratus-ss) [reply yt_dlp/extractor/gamedevtv.py:94]: This was the piece I was missing! I don't fully understand what's happening here. Thank you so much!!!

**Comment by stratus-ss** [comment]
> This is what I would have done

I've tested and implemented your suggestion. Thank you very much!

**Review by seproDev — APPROVED** [review]

File: yt_dlp/extractor/gamedevtv.py [hunk 6]
```diff
@@ -56,14 +56,16 @@ def _real_initialize(self):
             self.raise_login_required(
                 'This content is only available with purchase', method='password')
 
-    def _entries(self, data, course_id, course_info):
+    def _entries(self, data, course_id, course_info, lecture_id):
```
Comment (bashonly) [inline yt_dlp/extractor/gamedevtv.py:84]:
```suggestion
    def _entries(self, data, course_id, course_info, selected_lecture):
```

File: yt_dlp/extractor/gamedevtv.py [hunk 7]
```diff
@@ -56,14 +56,16 @@ def _real_initialize(self):
...
         for section in traverse_obj(data, ('sections', ..., {dict})):
             section_info = traverse_obj(section, {
                 'season_id': ('id', {str_or_none}),
                 'season': ('title', {str}),
                 'season_number': ('order', {int_or_none}),
             })
             for lecture in traverse_obj(section, ('lectures', lambda _, v: url_or_none(v['video']['playListUrl']))):
+                if lecture_id and str(lecture.get('id')) != lecture_id:
```
Comment (bashonly) [inline yt_dlp/extractor/gamedevtv.py:92]:
```suggestion
                if selected_lecture and str(lecture.get('id')) != selected_lecture:
```

**Review by bashonly — APPROVED** [review]

**Pull request merged**